Product Subscriptions: Add expire after view

selanthiraiyan opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11175
<!-- Id number of the GitHub issue this PR addresses. -->

## Description

- Adds a new row in product details to view the Expire after info of a subscription product.
- Updates the "Free trial" screen UI to use SwiftUI form

## Testing instructions

**Prerequisites**
- Enable `subscriptionProducts` feature flag by returning `true` from [this line](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Experiments/Experiments/DefaultFeatureFlagService.swift#L90) 
- Launch the app and login 

**Free trial**
- Navigate to products
- Tap on a simple subscription product (Create one from `wpadmin` if needed)
- Check that the Expire after info is displayed in the product details screen
- Tap on the Expire after row
- Validate that the Expire after has the values based on the subscription's billing interval. (Use `wpadmin` behaviour to cross validate) p1700121516912349-slack-C03L1NF1EA3
- Ensure that you can successfully select a new "Expire after" value and save the product.

**Free trial**
- Navigate to products
- Tap on a simple subscription product (Create one from `wpadmin` if needed)
- Check that the Free trial info is displayed in the product details screen
- Tap on the Free trial row
- Smoke test the screen to ensure SwiftUI `Form` based UI works as expected.

## Screenshots
| Free trial | Expire after |
|--------|--------|
| ![Simulator Screenshot - iPhone 14 Pro - 2023-11-20 at 11 00 05](https://github.com/woocommerce/woocommerce-ios/assets/524475/9c0b4b95-a2ff-43e4-92e9-963b0e8f328f) | ![Simulator Screenshot - iPhone 14 Pro - 2023-11-20 at 11 00 10](https://github.com/woocommerce/woocommerce-ios/assets/524475/ed619e2c-c10c-45f5-9fad-65db9bb79486) | 

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Add calendar clock image.

selanthiraiyan committed:
Add method to update product/variation with expiry info.

selanthiraiyan committed:
Sets "Expire after" to "0" if the subscription period or interval is changed

selanthiraiyan committed:
Add view to show/edit subscription trial info.

selanthiraiyan committed:
Show subscription expiry row in product form.

selanthiraiyan committed:
Present subscription expiry screen from product form.

selanthiraiyan committed:
Add unit tests for `SubscriptionExpiryViewModel`.

selanthiraiyan committed:
Unit test that "Expire after" row is displayed.

selanthiraiyan committed:
Unit test that "Expire after" row is displayed with expected values.

selanthiraiyan committed:
Unit test that subscription length is reset when period/interval changes.

selanthiraiyan committed:
Unit test the subscription length is updated properly.

selanthiraiyan committed:
Resolve conflicts after merging changes from trunk.

selanthiraiyan committed:
Update UI to follow wheel style in picker.

selanthiraiyan committed:
Make height and padding configurable.

selanthiraiyan committed:
Make height and padding configurable.

selanthiraiyan committed:
Remove unused layout constants.

selanthiraiyan committed:
Refactor subscription trial view to use form.

selanthiraiyan committed:
Merge branch 'trunk' into feat/11175-expiry

selanthiraiyan committed:
Enable `Done` button only when selection changes.

selanthiraiyan committed:
Fix checking trial length and trial period for expiry in unit tests.

selanthiraiyan committed:
Merge branch 'trunk' into feat/11175-expiry

hafizrahman commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/DefaultProductFormTableViewModel.swift:
```diff
@@ -692,6 +710,29 @@ extension DefaultProductFormTableViewModel {
+                                                               value: "Expire after",
+                                                               comment: "Title for Subscription expiry row in the product form screen.")
+        static let neverExpire = NSLocalizedString("defaultProductFormTableViewModel.neverExpire",
+                                                   value: "Never expire",
```
np: I understand this follows Core but I think "Never expires" is more grammatically correct (ref: https://oneminuteenglish.org/en/verb-tense-after-never/ "Never in present tense")

selanthiraiyan replied:
Good point. Updated the text in 887a46459098d564bd418e024a73f3ad96d087d7

hafizrahman commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductVariationFormViewModel.swift:
```diff
@@ -342,6 +353,15 @@ extension ProductVariationFormViewModel {
     }
+
+    func updateSubscriptionExpirySettings(length: String) {
+        // TODO: 11090 - Check this while implementing variation form
```
❓  What needs to be checked here?

selanthiraiyan replied:
We need to test that the values of the variation are updated properly. I can see that you have already checked that in your PR https://github.com/woocommerce/woocommerce-ios/pull/11221.

Removed TODOs in 23c7ed2be51bbca6d006665534ebc59dd227dd0e.

hafizrahman commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Subscription/SubscriptionExpiryViewModelTests.swift:
```diff
@@ -0,0 +1,129 @@
+        let viewModel = SubscriptionExpiryViewModel(subscription: subscription) { _, _ in }
+
+        // Then
+        XCTAssertEqual(viewModel.selectedLength.title, "Never expire")
```
np, possibly same situation (here and the ones below) as in this comment: https://github.com/woocommerce/woocommerce-ios/pull/11173#discussion_r1396615549 ?

selanthiraiyan replied:
Good catch. Also fixed in 887a46459098d564bd418e024a73f3ad96d087d7

hafizrahman reviewed (approved):
Works well in my test and the code also looks great, consistent with the free trial section work before it. I left some comment and question but they're non-blocking.

selanthiraiyan committed:
Update Never expires text.

selanthiraiyan committed:
Remove TODO checks.

selanthiraiyan commented:
Thanks for the review, @hafizrahman! I have addressed your comments. I am enabling auto merge now to unblock other PRs. Please feel to add more comments if needed. I will address any new comments in a future PR.

selanthiraiyan committed:
Don't update "Never expire" text in legacy subscription row.

itsmeichigo commented:
Hi there! I'm not sure if this is a regression issue from updating the free trial screen to use `Form`, but I saw this minor issue with the extra leading space of the error message:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/fd2698c5-73b1-47e3-96cf-750c454240af" width=320 />

I'll add this as a minor issue to fix for #11090.

selanthiraiyan commented:
Nice catch, Huong! Thanks for reporting.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductVariationFormViewModel.swift:
```diff
@@ -228,7 +228,18 @@ extension ProductVariationFormViewModel {
+        // Sets "Expire after" to "0" (i.e. Never expire)
+        // if the subscription period or interval is changed
+        let subscriptionLength: String? = {
+            if productVariation.subscription?.period != subscriptionPeriod || productVariation.subscription?.periodInterval != subscriptionPeriodInterval {
+                return "0"
+            } else {
```
Just wondering: I cannot get my head around this logic even with the comment above. @selanthiraiyan could you please explain again why we should set the length to 0 in this case?

selanthiraiyan replied:
- When the period is changed, the expiry date options (in drop down) will change as well. 
- To avoid selecting a non-matching expiry date value we are resetting the expiry date back to "Never expire".
- The user will have to select a valid expiry date. The expiry date options will be calculated based on the entered period. 

Here is the web behaviour. 

https://github.com/woocommerce/woocommerce-ios/assets/524475/db8c5aee-187d-4444-a1b0-2e9c9ef53d47

Here is the discussion in which @JorgeMucientes shared about this behaviour - p1700124102155059/1700121516.912349-slack-C03L1NF1EA3

Kindly let me know if you have any questions.

itsmeichigo replied:
Oh, this makes sense! It was hard to understand because the two fields are on separate screens, making it hard to connect them.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/ValidationErrorRow.swift:
```diff
@@ -3,7 +3,17 @@ import SwiftUI
-    let errorMessage: String
+    private let errorMessage: String
+    private let minHeight: CGFloat
+    private let horizontalPadding: CGFloat
```
Looks like this padding is not used in the view, which is why the error text has the extra padding as I reported yesterday.

selanthiraiyan replied:
Fixed this in https://github.com/woocommerce/woocommerce-ios/pull/11310
